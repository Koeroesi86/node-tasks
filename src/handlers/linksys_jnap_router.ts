import axios from 'axios';

export async function callJNAPRouter({
  router = '',
  user = '',
  password = '',
  action = '',
  data = {},
}) {
  const baseURL = `http://${router}/`;
  const response = await axios.request({
    baseURL,
    url: '/JNAP/',
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'X-JNAP-Authorization': `Basic ${Buffer.from(
        `${user}:${password}`,
        'utf-8'
      ).toString('base64')}`,
      'X-JNAP-Action': `http://linksys.com/jnap/${action}`,
    },
    data,
  });

  if (response.data.result !== 'OK') {
    throw Object.assign(new Error(response.data.result), response.data);
  }

  return response.data;
}
