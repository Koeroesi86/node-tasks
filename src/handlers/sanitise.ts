import moxios from 'moxios';

export const sanitiseMoxios = (
  request: ReturnType<typeof moxios.requests.at>
) => ({
  baseUrl: request.config.baseURL,
  url: request.config.url,
  method: request.config.method,
  headers: { ...request.config.headers },
  data: request.config.data,
});
