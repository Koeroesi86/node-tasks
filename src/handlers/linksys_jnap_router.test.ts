import moxios from 'moxios';
import { callJNAPRouter } from './linksys_jnap_router';
import { sanitiseMoxios } from './sanitise';

describe('callJNAPRouter', () => {
  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('calls axios', async () => {
    const router = 'example-router';
    moxios.stubOnce('POST', `/JNAP/`, {
      status: 200,
      response: { result: 'OK' },
    });

    await callJNAPRouter({
      router,
      user: 'example-user',
      password: 'example-password',
      action: 'example-password',
      data: { example: 'data' },
    });

    expect(moxios.requests.count()).toEqual(1);
    expect(sanitiseMoxios(moxios.requests.at(0))).toMatchSnapshot();
  });
});
