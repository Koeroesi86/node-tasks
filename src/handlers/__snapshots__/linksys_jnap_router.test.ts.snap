// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`callJNAPRouter calls axios 1`] = `
{
  "baseUrl": "http://example-router/",
  "data": "{"example":"data"}",
  "headers": {
    "Accept": "application/json",
    "Content-Type": "application/json",
    "X-JNAP-Action": "http://linksys.com/jnap/example-password",
    "X-JNAP-Authorization": "Basic ZXhhbXBsZS11c2VyOmV4YW1wbGUtcGFzc3dvcmQ=",
  },
  "method": "post",
  "url": "/JNAP/",
}
`;
